File: python/DTW.py
import numpy as np


def distance(s1, s2):
    n = len(s1) + 1
    m = len(s2) + 1

    d = np.zeros((n, m))

    for i in range(n):
        d[i, 0] = np.inf
    for i in range(m):
        d[0, i] = np.inf

    d[0][0] = 0

    for i in range(1, n):
        for j in range(1, m):
            d[i][j] = abs(s1[i - 1] - s2[j - 1]) + min(d[i - 1][j - 1], d[i - 1][j], d[i][j - 1])

    return d[n - 1][m - 1]

File: matlab/dtw.m
function [distance] = dtw(s1,s2)

n = length(s1) + 1;
m = length(s2) + 1;

d = zeros(n, m, 'double');

for i = 1 : n
    d(i,1) = inf;
end


for i = 1 : m
    d(1,i) = inf;
end

d(1,1) = 0;


for i = 2 : n
    for j = 2 : m
        d(i,j) = abs(s1(i-1) - s2(j-1)) ... 
            + min(d(i-1,j), min(d(i,j-1),d(i-1,j-1)));
    end
end

distance = d(i,j);
